zip-0400.rst

```diff
@@ -0,0 +1,174 @@
+* If the wallet is created with HD support, it will have additional properties ``hdseed`` and ``hdchain`` that will be saved.
+* ``version``, ``minversion``, ``witnesscachesize`` and ``bestblock`` properties are added. This are settings and state information, the ``bestblock`` property is a good example of the database being populated that is happening without any user interaction but it will just update as the best block in the current chain changes.
+
+At any time after the database is created new properties can be added as the wallet users do actions. For example, if the user creates a new sapling address with the rpc command ``getznewaddress()`` then new records with properties `sapzkey` and `sapzkeymeta` will be added to the database.
+
+Database changes do not happen immediately but their are flushed in its own thread by ``ThreadFlushWalletDB()`` function periodically to avoid overhead. The internal counter ``nWalletDBUpdated`` is increased each time a new write operation to the database is done, this is compared with the last flush in order to commit new stuff.
+
+When the node goes down for whatever reason the information in the database will persist in the disk, the next time the node start, the software will detect the database file, read from there and add the values into memory structures that will guarantee wallet functionality.
+
+Transactions
+^^^^^^^^^^^^
+
+The database wallet will not save all the transactions that are happening in the blockchain however it will save all transactions where wallet keys are involved. This is needed for example to get balances, in order to do this the wallet must have all the transactions related to a key to compute the final value of coin available in the derived address.
+
+The ``tx`` property will hold this transactions related data with the hash as the key and the full transaction as the value.
+
+Wallet state and transaction reordering
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There is a small class ``CWalletScanState`` that will hold some information about the current state of the wallet like the number of keys of all types, if the wallet is encrypted and others. The complete list of variable members is::
+
+    class CWalletScanState {
+    public:
+        unsigned int nKeys;
+        unsigned int nCKeys;
+        unsigned int nKeyMeta;
+        unsigned int nZKeys;
+        unsigned int nCZKeys;
+        unsigned int nZKeyMeta;
+        unsigned int nSapZAddrs;
+        bool fIsEncrypted;
+        bool fAnyUnordered;
+        int nFileVersion;
+        vector<uint256> vWalletUpgrade;
+    };
+
+If transaction order has changed(orphan blocks, others) then ``fAnyUnordered`` will be true and the wallet code will rescan all transaction again. Transactions may change their order by several reasons, this is handled by wallet code but needs to be reflected by database code as well thus the rescan. 
```

Comment:
This is implementation detail of zcashd, do we need this to understand the wallet db format?

Reply:
If someone is building a new wallet i think it can be important to know the implementation should handle a possible modification in the order in the transactions. I agree the code is implementation detail and can be removed. I will see how i word it and send a commit soon.

Reply:
https://github.com/zcash/zips/pull/372/commits/96f0747e178450cac695ace04cb313886991817e